legion/engine/llri-dx/source.cpp

```diff
@@ -9,88 +10,143 @@ namespace legion::graphics::llri
+                llri::destroyInstance(output);
+                return internal::mapHRESULT(factoryCreateResult);
             }
+
+            //Store factory and return result
+            output->m_ptr = factory;
+            *instance = output;
+            return result::Success;
         }
 
-        //Attempt to create factory
-        IDXGIFactory* factory = nullptr;
-        const HRESULT factoryCreateResult = CreateDXGIFactory2(factoryFlags, IID_PPV_ARGS(&factory));
-        if (FAILED(factoryCreateResult))
+        void impl_destroyInstance(Instance* instance)
         {
-            destroyInstance(output);
-            return internal::mapHRESULT(factoryCreateResult);
-        }
+            if (!instance)
+                return;
 
-        //Store factory and return result
-        output->m_ptr = factory;
-        *instance = output;
-        return result::Success;
-    }
+            for (auto& [ptr, adapter] : instance->m_cachedAdapters)
+                delete adapter;
 
-    void destroyInstance(Instance* instance)
-    {
-        if (!instance)
-            return;
+            if (instance->m_debugAPI)
+                static_cast<ID3D12Debug*>(instance->m_debugAPI)->Release();
 
-        for (auto& [ptr, adapter] : instance->m_cachedAdapters)
-            delete adapter;
+            if (instance->m_debugGPU)
+                static_cast<ID3D12Debug1*>(instance->m_debugGPU)->Release();
 
-        if (instance->m_debugAPI)
-            static_cast<ID3D12Debug*>(instance->m_debugAPI)->Release();
+            delete instance;
+        }
 
-        if (instance->m_debugGPU)
-            static_cast<ID3D12Debug1*>(instance->m_debugGPU)->Release();
+        void impl_pollAPIMessages(const validation_callback_desc& validation, void* messenger)
+        {
+            if (messenger != nullptr && messenger != (void*)1)
+            {
+                ID3D12InfoQueue* iq = static_cast<ID3D12InfoQueue*>(messenger);
+                const auto numMsg = iq->GetNumStoredMessages();
+                
+                for (UINT64 i = 0; i < numMsg; ++i)
+                {
+                    SIZE_T messageLength = 0;
+                    iq->GetMessage(i, NULL, &messageLength);
+                    
+                    D3D12_MESSAGE* pMessage = (D3D12_MESSAGE*)malloc(messageLength);
+                    iq->GetMessage(i, pMessage, &messageLength);
+                    validation(internal::mapSeverity(pMessage->Severity), validation_callback_source::InternalAPI, pMessage->pDescription);
+                    
+                    free(pMessage);
+                }
 
-        delete instance;
+                iq->ClearStoredMessages();
+            }
+        }
     }
 
-    result Instance::enumerateAdapters(std::vector<Adapter*>* adapters)
+    result Instance::impl_enumerateAdapters(std::vector<Adapter*>* adapters)
     {
-        if (adapters == nullptr)
-            return result::ErrorInvalidUsage;
-
         adapters->clear();
```

Comment:
I understand the usage here, but honestly anywhere else using `clear()` on a vector of pointers without making sure they're deleted would make me VERY nervous. This to me screams for causing leaks somewhere at some point in time.

Reply:
I understand your concern but these are non-owning pointers so that is fortunately not an issue. The clear is only really there in case someone calls enumerateAdapters() multiple times and accidentally fills duplicates into their vector

Reply:
yeah i get that, but that's the reason why in legion core we use `pointer<SomeType>` which is a trivial wrapper around a non owning pointer. the only thing it does is it removes the ability to use `delete`, and it get's completely removed at compiletime.

i get that this might not be an option for LLRI but I just wanted to mention it.